### ISSUE TITLE
Is pattern evaluation order guaranteed?

### ISSUE THREAD

**Issue by zachs18** [rust-lang/unsafe-code-guidelines#540]
cc https://github.com/rust-lang/reference/issues/1665

For most patterns and in safe code, "evaluation"(/matching?) order of subpatterns does not matter, but there is (that I can think of) one instance on stable where pattern evaluation order matters: matching on a struct with a tag and a `union` field (and similar situations).

The Reference section on `union`s does mention [pattern matching](https://doc.rust-lang.org/reference/items/unions.html#pattern-matching-on-unions), but does not say anything about pattern evaluation order. It gives an example of pattern-matching on a manual tagged union, though pattern evaluation order does not matter for the example given[^2]. In a [slightly different example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e18ddd85d732df0bea8f7bf59c7f2ed6), however, the field order does matter:

<details> <summary>my original example</summary>

```rs
#[derive(Clone, Copy)]
enum Tag { A, B, }

#[derive(Clone, Copy)]
#[repr(C)]
union Value {
    a: u32,
    b: u8, // note that b is smaller than a
}

/// Assume that if tag == Tag::A, then val.a is valid, and if tag == Tag::B, then tag.b is valid.
#[derive(Clone, Copy)]
struct Tagged {
    tag: Tag,
    val: Value,
}

unsafe fn tag_first(v: Tagged) -> bool {
    match v {
        // fine under miri with tag == B, sees that `tag != A` and skips the arm
        Tagged { tag: Tag::A, val: Value { a: 0 } } => true,
        _ => false,
    }
}
unsafe fn val_first(v: Tagged) -> bool {
    match v {
        // error under miri with tag == B, since it reads the padding bytes after `Value::b`
        Tagged { val: Value { a: 0 }, tag: Tag::A } => true,
        _ => false,
    }
}

fn main() {
    let v = Tagged {
        tag: Tag::B,
        val: Value { b: 0 },
    };
    unsafe {
        tag_first(v);
        val_first(v);
    }
}
```

</details>

a simpler but basically the same example

```rs
fn main() {
    union Union { value: u8, _empty: () }
    struct MyOption { tag: u8, value: Union } // assume tag == 1 means value.value is valid
    let foo = MyOption { tag: 0, value: Union { _empty: () } };
    unsafe {
        match foo {
          // currently fine under Miri, since `tag` is mentioned first
          MyOption { tag: 1, value: Union { value: 0 } } => true,
          _ => false,
        };
        match foo {
          // currently this is UB under Miri if value is `_empty`/uninit, regardless of the tag field
          MyOption { value: Union { value: 0 }, tag: 1 } => true,
          _ => false,
        };
    }
}
```

For unstable code, I suppose `deref_patterns` might also make it important to document pattern evaluation order, or maybe that feature is/will be restricted enough for it not to matter. Depending on the resolution of https://github.com/rust-lang/unsafe-code-guidelines/issues/412 pattern evaluation order might be important if matching on references-to-references-to-invalid-data ([miri example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4b4b5b02c81c1cb8b2f7640090327e78))?

I'm not sure if this is fully the intended behavior[^1], or if it is intended, how best to document it.

[^1]: Alternately, instead of documenting pattern evaluation order, it could be specified that if any (union) field used in a pattern match is invalid/uninitialized, then the whole arm is UB, regardless of the order the fields were written in the pattern.

[^2]: in that example, the union field is fully-initailized either way, or UB happens regardless of pattern evaluation order

**Comment by theemathas** [comment]
Here's a version of the issue which doesn't use union patterns:

```rust
#[repr(C)]
struct Foo {
    flag: bool,
    payload: u8,
}

#[repr(C)]
struct Bar {
    flag: bool,
    payload: bool,
}

fn main() {
    let data = Foo {
        flag: false,
        payload: 2,
    };
    let bar_ref: &Bar = unsafe { &*(&raw const data).cast() };
    println!("first match...");
    match *bar_ref {
        Bar {
            flag: true,
            payload: false,
        } => {}
        _ => {}
    }
    println!("second match...");
    match *bar_ref {  // UB occurs here
        Bar {
            payload: false,
            flag: true,
        } => {}
        _ => {}
    }
}
```

```
error: Undefined Behavior: interpreting an invalid 8-bit value as a bool: 0x02
  --> src/main.rs:28:5
   |
28 |     match *bar_ref {  // UB occurs here
   |     ^^^^^^^^^^^^^^ Undefined Behavior occurred here
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: inside `main` at src/main.rs:28:5: 28:19

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
```

**Comment by RalfJung** [comment]
Could you add some commentary so we don't have to reverse engineer your code?

**Comment by theemathas** [comment]
@RalfJung In my code, `bar_ref` is a reference to a `Bar` with `flag = false`, and an invalid value for `payload`. This is fine, assuming that reference validity is shallow.

In the first `match`, we first check if `flag == true`, and since it's not, we immediately exit the pattern-matching. This happens before we get around to discovering that `payload` has an invalid value. Therefore, no UB happens.

In the second `match`, we first check if `payload == false`. Since `payload` has an invalid value, this causes UB.

The only difference between the two `match`es is the ordering of the fields in the pattern.

**Comment by theemathas** [comment]
Based on the fact that this code doesn't have UB according to Miri: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7b57e415bccac506d4324dd2aab6be7f

I'm guessing that the matching thing currently works as follows:
1. It checks whether the value matches the pattern, one subpattern by subpattern, in a fixed order.
  * It doesn't matter which order an `ident @ pat` subpattern is checked, since the ident part is irrefutable
  * This is an opsem read for "non-composite" subpatterns other than identifier patterns and underscore patterns.
2. Then, it binds to all bindings in the whole pattern, one subpattern by subpattern, in a fixed order.
  * In an `ident @ pat` subpattern, it binds to the pat before binding to the ident.
  * This is what borrowck cares about.
  * This is an opsem read for "non-composite" subpatterns other than underscore patterns.

Is there any test case that contradicts this mental model?

See also [#t-lang > Evaluation-order of patterns (&amp;miri-observable consequences)](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/Evaluation-order.20of.20patterns.20.28.26miri-observable.20consequences.29/with/606385150)

**Comment by theemathas** [comment]
Or patterns also act weird. This code snippet is by @steffahn
```rust
fn main() {
    let data: [u8; 3] = [0, 2, 0];
    let ptr: *const [bool; 3] = (&raw const data).cast();
    unsafe {
        if let [false, false | true, true] = *ptr {} // no UB
        if let [false, false | true, false] = *ptr {} // yes UB
    }
}
```

It seems like or subpatterns are checked after all the other subpatterns? This makes no sense whatsoever.

**Comment by theemathas** [comment]
I found a surprising interaction between or-patterns and union patterns. I've filed this as https://github.com/rust-lang/rust/issues/158387.

**Comment by theemathas** [comment]
@steffahn found that one match arm's contents can affect the match ordering in another arm.

This makes negative sense.

```rust
#[repr(u8)]
pub enum Enum {
    A,
    B,
    C,
    D,
}

#[repr(C)]
struct S(Enum, bool);

fn main() {
    let data: [u8; 2] = [42, 0];
    let ptr: *const S = (&raw const data).cast();
    unsafe {
        match *ptr {
            S(Enum::A | Enum::B, true) => (), // no UB
            // Deleting the above arm causes the below arm to have UB
            S(Enum::C, true) => (), // no UB – even though this one is a `|`-less pattern
            S(Enum::D, true) => (), // no UB – even though this one is a `|`-less pattern
            _ => println!("nothing matched"), // this prints
        }
    }
}
```

**Comment by Nadrieril** [comment]
The current guarantees for pattern-matching opsem are nonexistent. For that reason the union patterns in the OP are unsound and the weird pattern ordering behavior you noticed is allowed (and in fact purposeful).

My working model is to specify all this as a desugaring that non-deterministically choses in which order to test independent subconditions. The current behavior is a valid refinement of that model; we could want to strengthen it but I suspect the optimization potential of keeping it unspecified is quite high.

EDIT: well, I just learned that apparently the Reference does guarantee this union pattern behavior (https://github.com/rust-lang/rust/issues/158387) x) seems that no one told us, the match lowering code is very much written with reorderability in mind.

**Comment by RalfJung** [comment]
> EDIT: well, I just learned that apparently the Reference does guarantee this union pattern behavior (https://github.com/rust-lang/rust/issues/158387) x) seems that no one told us, the match lowering code is very much written with reorderability in mind.

Yeah I was about to say, what you said contradicts the Reference. I was not aware of this text in the Reference either. I filed an issue in that repo as well: https://github.com/rust-lang/reference/issues/2298.
